File: node/test/webhook.test.ts
import { describe, expect, it } from 'vitest';
import {
  EnvelopeCodec,
  PlutusWebhookError,
  WebhookHandler,
  assertWebhookEnvelope,
  buildAad,
  buildWebhookCanonicalString,
  loadPrivateKey,
  signCanonicalString,
  verifyCanonicalSignature,
  webhookBodyDigestBase64,
} from '../src/index.js';
import { key, vectors } from './vectors.js';

const platformAuth = key('platform_auth');
const merchantEnc = key('merchant_enc');

function handler(overrides: Record<string, unknown> = {}): WebhookHandler {
  return new WebhookHandler({
    platformAuthPublicKey: platformAuth.publicKeyPem,
    merchantEncPrivateKey: merchantEnc.privateKeyPem,
    apiKeyBizId: 'apk_vector_0001',
    ...overrides,
  });
}

describe('webhook 向量组', () => {
  it('向量数量为 2', () => {
    expect(vectors.vectors.webhook).toHaveLength(2);
  });

  for (const vector of vectors.vectors.webhook) {
    describe(`${vector.id}: ${vector.description}`, () => {
      const rawBody = Buffer.from(vector.body, 'utf8');

      it('body 摘要是 Base64 而非 hex', () => {
        const digest = webhookBodyDigestBase64(rawBody);
        expect(digest).toBe(vector.bodyDigestBase64);
        expect(digest).toMatch(/=$|[A-Za-z0-9+/]$/);
        expect(digest).not.toMatch(/^[0-9a-f]{64}$/);
      });

      it('4 行签名规范串一致', () => {
        const canonical = buildWebhookCanonicalString(
          vector.headers['X-SlaunchX-Delivery-Id'] as string,
          vector.headers['X-SlaunchX-Event-Type'] as string,
          vector.headers['X-SlaunchX-Timestamp'] as string,
          webhookBodyDigestBase64(rawBody),
        );
        expect(canonical).toBe(vector.signatureCanonicalString);
        expect(canonical.split('\n')).toHaveLength(4);
      });

      it('用 platform_auth 公钥验签通过', () => {
        expect(
          verifyCanonicalSignature(
            vector.signatureCanonicalString,
            vector.headers['X-SlaunchX-Signature'] as string,
            platformAuth.publicKeyPem,
          ),
        ).toBe(true);
      });

      it('AAD 第二分量固定为 webhook,第四分量是 API Key 业务 ID', () => {
        expect(vector.aadComponents.routeTemplate).toBe('webhook');
        expect(vector.aadComponents.keyId).toBe(vector.headers['X-SlaunchX-Key-Id']);
        expect(vector.aadComponents.keyId).not.toMatch(/^SHA256:/);
        const aad = buildAad(vector.aadComponents);
        expect(aad.toString('utf8')).toBe(vector.aadString);
        expect(aad.toString('base64')).toBe(vector.aadBase64);
      });

      it('信封形状:有 envelopeVersion,无 encryptedPayload', () => {
        const envelope = assertWebhookEnvelope(JSON.parse(vector.body));
        expect(envelope.envelopeVersion).toBe(1);
        expect('encryptedPayload' in envelope).toBe(false);
        expect(envelope.keyFingerprint).toBe(merchantEnc.fingerprint);
      });

      it('WebhookHandler 验签并解密,明文与向量一致', () => {
        const delivery = handler().handle(rawBody, vector.headers);
        expect(delivery.plaintext).toBe(vector.expectedPlaintext);
        expect(delivery.headers.deliveryBizId).toBe(vector.headers['X-SlaunchX-Delivery-Id']);
        expect(delivery.payload.eventType).toBe(vector.headers['X-SlaunchX-Event-Type']);
        expect(delivery.payload.payloadSchemaVersion).toBe(1);
      });

      it('直接用信封与重建 AAD 解密,结果相同', () => {
        const envelope = assertWebhookEnvelope(JSON.parse(vector.body));
        expect(
          EnvelopeCodec.openText(envelope, {
            privateKey: merchantEnc.privateKeyPem,
            aad: vector.aadComponents,
          }),
        ).toBe(vector.expectedPlaintext);
      });

      it('金额字段保持十进制字符串,不被解析为浮点数', () => {
        const payload = JSON.parse(vector.expectedPlaintext) as Record<string, unknown>;
        const stack: unknown[] = [payload];
        while (stack.length > 0) {
          const node = stack.pop();
          if (!node || typeof node !== 'object') continue;
          for (const [k, v] of Object.entries(node as Record<string, unknown>)) {
            if (k === 'amount') expect(typeof v).toBe('string');
            if (v && typeof v === 'object') stack.push(v);
          }
        }
      });
    });
  }
});

describe('webhook 负向', () => {
  const vector = vectors.vectors.webhook[0] as NonNullable<(typeof vectors.vectors.webhook)[0]>;
  const rawBody = Buffer.from(vector.body, 'utf8');

  it('body 多一个字节即验签失败', () => {
    expect(() => handler().verify(Buffer.concat([rawBody, Buffer.from(' ')]), vector.headers)).toThrowError(
      /signature verification failed/,
    );
  });

  it('篡改 Delivery-Id 头即验签失败', () => {
    const headers = { ...vector.headers, 'X-SlaunchX-Delivery-Id': 'whd_forged_001' };
    expect(() => handler().verify(rawBody, headers)).toThrowError(/signature verification failed/);
  });

  it('篡改事件类型头即验签失败', () => {
    const headers = { ...vector.headers, 'X-SlaunchX-Event-Type': 'card.status' };
    expect(() => handler().verify(rawBody, headers)).toThrowError(/signature verification failed/);
  });

  it('缺少必需头时抛错', () => {
    const headers: Record<string, string> = { ...vector.headers };
    delete headers['X-SlaunchX-Signature'];
    expect(() => handler().verify(rawBody, headers)).toThrowError(/missing webhook header/);
  });

  it('Key-Id 与配置不符时抛错', () => {
    expect(() => handler({ apiKeyBizId: 'apk_other_0002' }).verify(rawBody, vector.headers)).toThrowError(
      /does not match the configured API key/,
    );
  });

  it('用 hex 摘要构造的规范串验签必然失败', () => {
    const hexDigest = Buffer.from(vector.bodyDigestBase64, 'base64').toString('hex');
    const canonical = buildWebhookCanonicalString(
      vector.aadComponents.requestId,
      vector.headers['X-SlaunchX-Event-Type'] as string,
      vector.aadComponents.timestamp,
      hexDigest,
    );
    expect(
      verifyCanonicalSignature(canonical, vector.headers['X-SlaunchX-Signature'] as string, platformAuth.publicKeyPem),
    ).toBe(false);
  });

  it('信封多出 encryptedPayload 字段即被拒绝(additionalProperties: false)', () => {
    const envelope = JSON.parse(vector.body) as Record<string, unknown>;
    envelope['encryptedPayload'] = envelope['ciphertext'];
    expect(() => assertWebhookEnvelope(envelope)).toThrowError(/unexpected field: encryptedPayload/);
  });

  it('envelopeVersion 不为 1 即被拒绝', () => {
    const envelope = JSON.parse(vector.body) as Record<string, unknown>;
    envelope['envelopeVersion'] = 2;
    expect(() => assertWebhookEnvelope(envelope)).toThrowError(/envelopeVersion must be exactly 1/);
  });

  it('AAD routeTemplate 用外部路径而非 webhook 字面量时解密失败', () => {
    const envelope = assertWebhookEnvelope(JSON.parse(vector.body));
    expect(() =>
      EnvelopeCodec.openText(envelope, {
        privateKey: merchantEnc.privateKeyPem,
        aad: { ...vector.aadComponents, routeTemplate: '/webhooks/card' },
      }),
    ).toThrowError(/rebuilt AAD does not match/);
  });

  it('AAD keyId 用指纹而非 API Key 业务 ID 时解密失败', () => {
    const envelope = assertWebhookEnvelope(JSON.parse(vector.body));
    expect(() =>
      EnvelopeCodec.openText(envelope, {
        privateKey: merchantEnc.privateKeyPem,
        aad: { ...vector.aadComponents, keyId: merchantEnc.fingerprint },
      }),
    ).toThrowError(/rebuilt AAD does not match/);
  });

  it('时间戳容差默认关闭,启用后可拒绝过旧投递', () => {
    expect(() => handler().verify(rawBody, vector.headers)).not.toThrow();
    const strict = handler({ timestampToleranceMs: 60_000, now: () => 1788142396408 + 120_000 });
    expect(() => strict.verify(rawBody, vector.headers)).toThrowError(/timestamp skew/);
    const lenient = handler({ timestampToleranceMs: 300_000, now: () => 1788142396408 + 120_000 });
    expect(() => lenient.verify(rawBody, vector.headers)).not.toThrow();
  });

  it('明文与传输头交叉校验失败时抛错', () => {
    const forgedHeaders = { ...vector.headers };
    const delivery = handler().handle(rawBody, vector.headers);
    expect(delivery.payload.deliveryBizId).toBe(forgedHeaders['X-SlaunchX-Delivery-Id']);

    const custom = new WebhookHandler({
      platformAuthPublicKey: platformAuth.publicKeyPem,
      merchantEncPrivateKey: merchantEnc.privateKeyPem,
    });
    // 用错配的 deliveryBizId 重新封装一份合法信封,但明文里的 deliveryBizId 保持原值。
    const mismatched = EnvelopeCodec.seal(vector.expectedPlaintext, {
      recipientPublicKey: merchantEnc.publicKeyPem,
      aad: {
        requestId: 'whd_mismatch_999',
        routeTemplate: 'webhook',
        timestamp: vector.aadComponents.timestamp,
        keyId: vector.aadComponents.keyId,
      },
      envelopeVersion: 1,
      includeEncryptedPayload: false,
    });
    const body = Buffer.from(JSON.stringify(mismatched), 'utf8');
    const canonical = buildWebhookCanonicalString(
      'whd_mismatch_999',
      vector.headers['X-SlaunchX-Event-Type'] as string,
      vector.aadComponents.timestamp,
      webhookBodyDigestBase64(body),
    );
    // 用 platform_auth 私钥重新签名,模拟一次合法投递但明文 deliveryBizId 不一致。
    const signature = signWith(canonical);
    expect(() =>
      custom.handle(body, {
        'X-SlaunchX-Delivery-Id': 'whd_mismatch_999',
        'X-SlaunchX-Event-Type': vector.headers['X-SlaunchX-Event-Type'] as string,
        'X-SlaunchX-Timestamp': vector.aadComponents.timestamp,
        'X-SlaunchX-Key-Id': vector.aadComponents.keyId,
        'X-SlaunchX-Signature': signature,
      }),
    ).toThrowError(PlutusWebhookError);
  });
});

/** 测试内联签名:仅用于构造负向用例,生产侧商户不持有 platform_auth 私钥。 */
function signWith(canonical: string): string {
  return signCanonicalString(canonical, loadPrivateKey(platformAuth.privateKeyPem));
}

File: go/webhook_test.go
package plutus

import (
	"bytes"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

func webhookHTTPHeader(vec webhookVector) http.Header {
	header := http.Header{}
	for name, value := range vec.Headers {
		header.Set(name, value)
	}
	return header
}

func TestWebhookVectors(t *testing.T) {
	for _, vec := range vectors(t).Vectors.Webhook {
		vec := vec
		t.Run(vec.ID, func(t *testing.T) {
			body := []byte(vec.Body)
			headers := WebhookHeadersFromHTTP(webhookHTTPHeader(vec))
			if headers.DeliveryID != vec.AADComponents.RequestID {
				t.Fatalf("delivery id = %q, want %q", headers.DeliveryID, vec.AADComponents.RequestID)
			}

			// 摘要是 Base64, 不是 hex。
			if got := WebhookBodyDigestBase64(body); got != vec.BodyDigestBase64 {
				t.Fatalf("body digest = %q, want %q", got, vec.BodyDigestBase64)
			}
			if got := headers.CanonicalString(body); got != vec.SignatureCanonicalString {
				t.Fatalf("canonical string = %q, want %q", got, vec.SignatureCanonicalString)
			}
			if lines := strings.Split(vec.SignatureCanonicalString, "\n"); len(lines) != 4 {
				t.Fatalf("webhook canonical string must have 4 lines, got %d", len(lines))
			}

			// AAD: routeTemplate 固定为 "webhook", keyId 是 API Key 业务 ID。
			aad := headers.AAD()
			if aad.RouteTemplate != WebhookRouteTemplate {
				t.Fatalf("route template = %q, want %q", aad.RouteTemplate, WebhookRouteTemplate)
			}
			if aad.String() != vec.AADString || aad.Base64() != vec.AADBase64 {
				t.Fatalf("aad = %q / %q, want %q / %q", aad.String(), aad.Base64(), vec.AADString, vec.AADBase64)
			}

			// 信封形状: 有 envelopeVersion, 无 encryptedPayload。
			env, err := ParseWebhookEnvelope(body)
			if err != nil {
				t.Fatalf("ParseWebhookEnvelope: %v", err)
			}
			if env.EnvelopeVersion == nil || *env.EnvelopeVersion != 1 {
				t.Fatal("webhook envelope must carry envelopeVersion = 1")
			}
			if env.EncryptedPayload != "" {
				t.Fatal("webhook envelope must not carry encryptedPayload")
			}
			if env.AAD != vec.AADBase64 {
				t.Fatalf("envelope aad echo = %q, want %q", env.AAD, vec.AADBase64)
			}

			receiver, err := NewWebhookReceiver(WebhookConfig{
				APIKey:                vec.Headers[HeaderWebhookKeyID],
				PlatformAuthPublicKey: publicKey(t, vec.SignatureVerificationKey),
				MerchantEncPrivateKey: privateKey(t, vec.DecryptionKey),
			})
			if err != nil {
				t.Fatalf("NewWebhookReceiver: %v", err)
			}
			if err := receiver.Verify(headers, body); err != nil {
				t.Fatalf("Verify: %v", err)
			}

			notification, err := receiver.Handle(headers, body)
			if err != nil {
				t.Fatalf("Handle: %v", err)
			}
			if string(notification.Plaintext) != vec.ExpectedPlaintext {
				t.Fatalf("plaintext = %q, want %q", notification.Plaintext, vec.ExpectedPlaintext)
			}
			if notification.Payload.DeliveryBizID != headers.DeliveryID ||
				notification.Payload.EventType != headers.EventType ||
				notification.Payload.PayloadSchemaVersion != 1 {
				t.Fatalf("payload cross-check failed: %+v", notification.Payload)
			}
			if notification.Payload.EventID == "" || notification.Payload.Resource.BizID == "" {
				t.Fatalf("payload is incomplete: %+v", notification.Payload)
			}

			// 负向: body 被篡改 (哪怕是尾随空白) 验签必须失败, 且不进入解密。
			if _, err := receiver.Handle(headers, append(append([]byte(nil), body...), ' ')); !errors.Is(err, ErrWebhookSignatureInvalid) {
				t.Errorf("tampered body: got %v, want ErrWebhookSignatureInvalid", err)
			}
			// 负向: 传输头分量被篡改。
			for name, mutate := range map[string]func(WebhookHeaders) WebhookHeaders{
				"deliveryId": func(h WebhookHeaders) WebhookHeaders { h.DeliveryID += "x"; return h },
				"eventType":  func(h WebhookHeaders) WebhookHeaders { h.EventType += "x"; return h },
				"timestamp":  func(h WebhookHeaders) WebhookHeaders { h.Timestamp = "1788142396409"; return h },
			} {
				if err := receiver.Verify(mutate(headers), body); !errors.Is(err, ErrWebhookSignatureInvalid) {
					t.Errorf("tampered header %s: got %v, want ErrWebhookSignatureInvalid", name, err)
				}
			}
			// 负向: keyId 与本商户 API Key 不符。
			wrongKey := headers
			wrongKey.KeyID = "apk_other_0002"
			if _, err := receiver.Handle(wrongKey, body); !errors.Is(err, ErrWebhookPayloadMismatch) {
				t.Errorf("wrong key id: got %v, want ErrWebhookPayloadMismatch", err)
			}
			// 负向: AAD 的 keyId 位被篡改后解密失败 (绕过头比对, 直接用错误 AAD 解封)。
			tampered := aad
			tampered.KeyID = "apk_other_0002"
			if _, err := Open(privateKey(t, vec.DecryptionKey), env, tampered); !errors.Is(err, ErrAADMismatch) {
				t.Errorf("tampered aad keyId: got %v, want ErrAADMismatch", err)
			}
		})
	}
}

func TestWebhookEnvelopeRejectsAPIChainShape(t *testing.T) {
	vec := vectors(t).Vectors.Webhook[0]
	var raw map[string]any
	if err := json.Unmarshal([]byte(vec.Body), &raw); err != nil {
		t.Fatalf("unmarshal body: %v", err)
	}

	// 多出 encryptedPayload 字段 (additionalProperties: false)。
	withExtra := map[string]any{}
	for k, v := range raw {
		withExtra[k] = v
	}
	withExtra["encryptedPayload"] = raw["ciphertext"]
	encoded, _ := json.Marshal(withExtra)
	if _, err := ParseWebhookEnvelope(encoded); !errors.Is(err, ErrEnvelopeInvalid) {
		t.Errorf("encryptedPayload must be rejected, got %v", err)
	}

	// 缺 envelopeVersion。
	withoutVersion := map[string]any{}
	for k, v := range raw {
		if k != "envelopeVersion" {
			withoutVersion[k] = v
		}
	}
	encoded, _ = json.Marshal(withoutVersion)
	if _, err := ParseWebhookEnvelope(encoded); !errors.Is(err, ErrEnvelopeInvalid) {
		t.Errorf("missing envelopeVersion must be rejected, got %v", err)
	}

	// envelopeVersion 不为 1。
	raw["envelopeVersion"] = 2
	encoded, _ = json.Marshal(raw)
	if _, err := ParseWebhookEnvelope(encoded); !errors.Is(err, ErrEnvelopeInvalid) {
		t.Errorf("envelopeVersion 2 must be rejected, got %v", err)
	}
}

func TestWebhookHandleRequest(t *testing.T) {
	vec := vectors(t).Vectors.Webhook[0]
	receiver, err := NewWebhookReceiver(WebhookConfig{
		APIKey:                vec.Headers[HeaderWebhookKeyID],
		PlatformAuthPublicKey: publicKey(t, vec.SignatureVerificationKey),
		MerchantEncPrivateKey: privateKey(t, vec.DecryptionKey),
	})
	if err != nil {
		t.Fatalf("NewWebhookReceiver: %v", err)
	}

	var captured *WebhookNotification
	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		notification, err := receiver.HandleRequest(r)
		if err != nil {
			http.Error(w, "rejected", http.StatusBadRequest)
			return
		}
		captured = notification
		w.WriteHeader(http.StatusNoContent)
	})

	req := httptest.NewRequest(http.MethodPost, "/webhooks/slaunchx", bytes.NewReader([]byte(vec.Body)))
	for name, value := range vec.Headers {
		req.Header.Set(name, value)
	}
	rec := httptest.NewRecorder()
	handler.ServeHTTP(rec, req)
	if rec.Code != http.StatusNoContent {
		t.Fatalf("status = %d, want %d", rec.Code, http.StatusNoContent)
	}
	if captured == nil || string(captured.Plaintext) != vec.ExpectedPlaintext {
		t.Fatal("handler did not decrypt the notification")
	}

	// 篡改的投递必须被拒绝。
	req = httptest.NewRequest(http.MethodPost, "/webhooks/slaunchx", strings.NewReader(vec.Body+" "))
	for name, value := range vec.Headers {
		req.Header.Set(name, value)
	}
	rec = httptest.NewRecorder()
	handler.ServeHTTP(rec, req)
	if rec.Code != http.StatusBadRequest {
		t.Fatalf("status = %d, want %d", rec.Code, http.StatusBadRequest)
	}
}

func TestWebhookTimestampTolerance(t *testing.T) {
	vec := vectors(t).Vectors.Webhook[0]
	deliveredAt := time.UnixMilli(1788142396408)
	receiver, err := NewWebhookReceiver(WebhookConfig{
		APIKey:                vec.Headers[HeaderWebhookKeyID],
		PlatformAuthPublicKey: publicKey(t, vec.SignatureVerificationKey),
		MerchantEncPrivateKey: privateKey(t, vec.DecryptionKey),
		TimestampTolerance:    5 * time.Minute,
		Now:                   func() time.Time { return deliveredAt.Add(time.Minute) },
	})
	if err != nil {
		t.Fatalf("NewWebhookReceiver: %v", err)
	}
	headers := WebhookHeadersFromHTTP(webhookHTTPHeader(vec))
	if _, err := receiver.Handle(headers, []byte(vec.Body)); err != nil {
		t.Fatalf("within tolerance: %v", err)
	}

	receiver, err = NewWebhookReceiver(WebhookConfig{
		APIKey:                vec.Headers[HeaderWebhookKeyID],
		PlatformAuthPublicKey: publicKey(t, vec.SignatureVerificationKey),
		MerchantEncPrivateKey: privateKey(t, vec.DecryptionKey),
		TimestampTolerance:    5 * time.Minute,
		Now:                   func() time.Time { return deliveredAt.Add(time.Hour) },
	})
	if err != nil {
		t.Fatalf("NewWebhookReceiver: %v", err)
	}
	if _, err := receiver.Handle(headers, []byte(vec.Body)); !errors.Is(err, ErrWebhookTimestampOutOfRange) {
		t.Fatalf("outside tolerance: got %v, want ErrWebhookTimestampOutOfRange", err)
	}
}

func TestMandatoryWebhookRecipient(t *testing.T) {
	cfg := WebhookConfig{PlatformAuthPublicKey: publicKey(t, "platform_auth"), MerchantEncPrivateKey: privateKey(t, "merchant_enc")}
	for _, id := range []string{"", " ", "\t\n"} {
		cfg.APIKey = id
		if _, err := NewWebhookReceiver(cfg); !errors.Is(err, ErrInvalidConfig) {
			t.Fatalf("missing local recipient accepted: %v", err)
		}
	}
	vec := vectors(t).Vectors.Webhook[0]
	headers := WebhookHeadersFromHTTP(webhookHTTPHeader(vec))
	body := []byte(vec.Body)
	cfg.APIKey = vec.Headers[HeaderWebhookKeyID]
	correct, err := NewWebhookReceiver(cfg)
	if err != nil {
		t.Fatal(err)
	}
	if _, err = correct.Handle(headers, body); err != nil {
		t.Fatal(err)
	}
	// Identical platform and encryption keys; only local API Key changes.
	cfg.APIKey = "apk_other_recipient"
	other, err := NewWebhookReceiver(cfg)
	if err != nil {
		t.Fatal(err)
	}
	if err = other.Verify(headers, body); err != nil {
		t.Fatal(err)
	}
	if _, err = other.Handle(headers, body); !errors.Is(err, ErrWebhookPayloadMismatch) {
		t.Fatalf("wrong recipient: %v", err)
	}
	if _, err = other.Handle(headers, append(append([]byte{}, body...), ' ')); !errors.Is(err, ErrWebhookSignatureInvalid) {
		t.Fatalf("signature must be checked first: %v", err)
	}
	headers.KeyID = cfg.APIKey
	if _, err = other.Handle(headers, body); err == nil {
		t.Fatal("rewriting header bypassed AAD")
	}
}
